Makefile

```diff
@@ -45,13 +58,37 @@ vet:
 # build targets
 #
 
-bin/%:
-	$(Q)echo "Building $@..."; \
-	$(GO_BUILD) -o $@ ./$(subst bin/,cmd/,$@)
+$(BINARIES): bin/%:
+	$(Q)echo "Building $@..."
+	$(Q)(cd cmd/$(*) && $(GO_BUILD) -o $(abspath $@) .)
+
+#
+# go module tidy and verify targets
+#
+.PHONY: mod-tidy $(CMD_MOD_TIDY_TARGETS) mod-tidy-root
+.PHONY: mod-verify $(CMD_MOD_VERIFY_TARGETS) mod-verify-root
+
+CMD_MOD_TIDY_TARGETS := mod-tidy-cdi mod-tidy-validate
+CMD_MOD_VERIFY_TARGETS := mod-verify-cdi mod-verify-validate
+
+mod-tidy-root:
+	$(Q)echo "Running $@..."; \
+	$(GO) mod tidy
+
+$(CMD_MOD_TIDY_TARGETS): mod-tidy-%: mod-tidy-root
+	$(Q)echo "Running $@... in $(abspath ./cmd/$(*))"; \
+	(cd $(abspath ./cmd/$(*)) && $(GO) mod tidy)
+
+mod-verify-root: mod-tidy-root
+	$(Q)echo "Running $@..."; \
+	$(GO) mod verify
 
-bin/cdi:
-	$(Q)echo "Building $@..."; \
-	cd cmd/cdi; $(GO_BUILD) -o $(abspath $@) .
+$(CMD_MOD_VERIFY_TARGETS): mod-verify-%: mod-tidy-% mod-verify-root
```

Comment:
@elezar Can you elaborate on this, please? Just wondering how it works and do we really need it.

Reply:
This is a pattern rule. The first part defines the targets and the second part defines the pattern to apply. It is also possible to define pattern-based dependencies.

In this case the `CMD_MOD_VERIFY_TARGETS` define the actual make targets that trigger this rule, the `mod-verify-%` applies a pattern to these targets and the `$(<)` make variable matches what every was captured by the `%` in the rule. The `mod-tidy-%` is a dependency where the `%` matches the `%` in the pattern rule.

It would be possible to expand these rules explicitly and I can do so if we feel strongly about not using these pattern rules.

Reply:
Thank you for the explanation. Now it's clear.